### ISSUE TITLE
ClusterWorkflowTemplate podMetadata are ignored with `workflowTemplateRef `

### ISSUE THREAD

**Issue by bygui86** [argoproj/argo-workflows#7091]
## Summary

I expected that creating a Workflow from a ClusterWorkflowTemplate, the resulting pods would inherit labels from `podMetadata`.
Those labels are mainly used to match specific NetworkPolicies granting access to internet and other services.

Argo Workflows version: 3.1.3

## Diagnostics

Either a workflow that reproduces the bug, or paste you whole workflow YAML, including status, something like:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ClusterWorkflowTemplate
metadata:
  name: go-pipeline
spec:
  serviceAccountName: cicd-pipeline
  activeDeadlineSeconds: 3600
  ttlStrategy:
    secondsAfterCompletion: 604800
  podMetadata:
    labels:
      domain: ci
      component: ci-executor-pod
      provider: http
...
```

Kubernetes provider:   GKE

executor: PNS

```bash
# Logs from the workflow controller:
kubectl logs -n argo deploy/workflow-controller | grep go-pipeline-***

time="2021-10-28T16:17:01.700Z" level=info msg="Processing workflow" namespace=cicd-pipelines workflow=go-pipeline-59vhk
time="2021-10-28T16:17:01.737Z" level=info msg="Updated phase  -> Running" namespace=cicd-pipelines workflow=go-pipeline-59vhk
time="2021-10-28T16:17:01.737Z" level=info msg="Creating pvc go-pipeline-59vhk-workdir" namespace=cicd-pipelines workflow=go-pipeline-59vhk
time="2021-10-28T16:17:01.746Z" level=info msg="DAG node go-pipeline-59vhk initialized Running" namespace=cicd-pipelines workflow=go-pipeline-59vhk
time="2021-10-28T16:17:01.747Z" level=info msg="All of node go-pipeline-59vhk.clone-code dependencies [] completed" namespace=cicd-pipelines workflow=go-pipeline-59vhk
time="2021-10-28T16:17:01.747Z" level=info msg="Pod node go-pipeline-59vhk-2560901167 initialized Pending" namespace=cicd-pipelines workflow=go-pipeline-59vhk
time="2021-10-28T16:17:01.781Z" level=info msg="Created pod: go-pipeline-59vhk.clone-code (go-pipeline-59vhk-2560901167)" namespace=cicd-pipelines workflow=go-pipeline-59vhk
time="2021-10-28T16:17:01.860Z" level=info msg="Workflow update successful" namespace=cicd-pipelines phase=Running resourceVersion=91309602 workflow=go-pipeline-59vhk
time="2021-10-28T16:18:01.778Z" level=info msg="Processing workflow" namespace=cicd-pipelines workflow=go-pipeline-59vhk
time="2021-10-28T16:18:01.779Z" level=info msg="Updating node go-pipeline-59vhk-2560901167 status Pending -> Running" namespace=cicd-pipelines workflow=go-pipeline-59vhk
time="2021-10-28T16:18:01.795Z" level=info msg="Workflow update successful" namespace=cicd-pipelines phase=Running resourceVersion=91310073 workflow=go-pipeline-59vhk


kubectl get pods --show-labels

NAME                                                     READY   STATUS      RESTARTS   AGE   LABELS
go-pipeline-59vhk-2641655615                             0/2     Pending     0          4s    workflows.argoproj.io/completed=false,workflows.argoproj.io/workflow=go-pipeline-59vhk
```

---

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by bygui86** [comment]
@sarabala1979 the `ClusterWorkflowTemplate` is referenced by an ArgoEvent `Sensor` this way:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Sensor
metadata:
  name: http-go
spec:
  replicas: 1
  template:
    serviceAccountName: ***
    metadata:
      labels:
        # ...
    securityContext:
      runAsUser: 1001
  dependencies:
    - name: http-go-dependency
      eventSourceName: ***
      eventName: http-go-event
  triggers:
    - template:
        conditions: http-go-dependency
        name: http-go-event-trigger
        k8s:
          group: argoproj.io
          version: v1alpha1
          resource: workflows
          operation: create
          source:
            resource:
              apiVersion: argoproj.io/v1alpha1
              kind: Workflow
              metadata:
                generateName:
                namespace: ***
                labels:
                  repo:
                  revision:
              spec:
                podMetadata:
                  labels:
                    domain: ci
                    component: ci-executor-pod
                    provider: http
                arguments:
                  parameters:
                    - name: repo_url
                      value: dummy-repo-url
                    - name: context_path
                      value: dummy-context-path
                    - name: revision
                      value: master
                    - name: container_image_tag_revision
                      value: dummy-container-image-tag-revision
                    - name: checkout_sha
                      value: http-trigger
                    - name: base_path
                      value: /go/src/gitlab.com/
                    - name: dockerfile_path
                      value: Dockerfile
                    - name: container_image_registry
                      value: eu.gcr.io/ceiba-platform-cicd/
                entrypoint: go-pipeline
                workflowTemplateRef:
                  name: go-pipeline
                  clusterScope: true
          parameters:
            # ... parameters mapping from http request ...
```

If I don't specify `podMetadata` in the Sensor, none of my custom labels are set on the workflow pods.

ArgoEvents version:   1.4.0

**Comment by sarabala1979** [comment]
@bygui86 Do you have podMetaData in clusterWorkflowTemplate?  Is only `podMetaData` in the sensor get applied?

**Comment by bygui86** [comment]
@sarabala1979 yes exactly, I ended up setting `podMetadata` on both `Sensor` and `ClusterWorkflowTemplate` otherwise it does not work.

I tried setting `podMetadata` only on ClusterWorkflowTemplate, together with other config like `serviceAccountName`, `activeDeadlineSeconds`, `ttlStrategy` and others, but it seems that right `podMetadata` gets ignored.

I tried setting `podMetadata` on the Sensor and it works, I see those labels on the pod.

I checked the documentation and the examples many times, but I didn't find anything special.

**Comment by sarabala1979** [comment]
@bygui86 can you try the latest version 3.1.14 or 3.2.3?

**Comment by bygui86** [comment]
@sarabala1979 thanks for the suggestion. At the moment we cannot upgrade as many things are in progress, but I hope to be able to try within couple of weeks.
Keep you posted on this!

**Comment by amaloun** [comment]
@sarabala1979 I have the same issue with `WorkflowTemplate` using the latest version `v3.2.3` , I created A `workflowTemplate` with labels in `podMetadata`, this labels should be applied to the workflow pods. But they were ignored. 


## Diagnostics


Here is a `WorkflowTemplate` and a `Workflow` to reproduce the bug.

- WorkflowTemplate
```yaml
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate                  
metadata:
  name: hello-world-template    
  namespace: argo
spec:
  entrypoint: main
  podMetadata:
    labels:
        app: whalesay
        tier: demo          
  templates:
  - name: main              
    container:
      image: docker/whalesay
      command: [cowsay]
      args: ["hello world"]
      resources:                
        limits:
          memory: 32Mi
          cpu: 100m
```
- Workflow 
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: hello-world-
  namespace: argo
spec:
  entrypoint: main
  templates:
  - name: main
    steps:
    - - name: main
        templateRef:
            name: hello-world-template
            template: main
```
What Kubernetes provider are you using? I'm running Argo on a Gke cluster with following version of k8s : `Kubernetes v1.21.5`

What executor are you running? `Emissary`

```bash
# description of the created pod  on k8s
kubectl -n argo describe pod hello-world-rwrdj-3091738535
Name:         hello-world-rwrdj-3091738535
Namespace:    argo
Priority:     0
Node:         gke-main-cluster-default-202103121115-5d220df4-qdja/10.100.0.13
Start Time:   Tue, 09 Nov 2021 14:59:45 +0100
Labels:       workflows.argoproj.io/completed=false
              workflows.argoproj.io/workflow=hello-world-rwrdj
Annotations:  cni.projectcalico.org/containerID: a87c79ed09509eb5e5007c0a1b0188b394d6b64a0845a06068bf99ebc801a927
              cni.projectcalico.org/podIP: 10.92.9.174/32
              cni.projectcalico.org/podIPs: 10.92.9.174/32
              workflows.argoproj.io/node-id: hello-world-rwrdj-3091738535
              workflows.argoproj.io/node-name: hello-world-rwrdj[0].main
...
...
```
I was expecting the following labels to be added to the pod labels
```bash
    labels:
        app: whalesay
        tier: demo 
```

**Comment by guillaumBrisard** [comment]
@amaloun I observe exactly the same thing with argo 3.2.3, emissary and kubernetes 1.19.6

**Comment by sarabala1979** [comment]
@amaloun above. your usecase is totally different than @bygui86 usecase.  For `templateRef` sceanior can't be refer the `spec` level properties. if you define the podmetadata under template it will work.

@bygui86  I tried my env. it is working fine.

**Comment by bygui86** [comment]
@sarabala1979 thanks for the update. May I ask you which version? A new one or just a different config?

**Comment by guillaumBrisard** [comment]
> @amaloun above. your usecase is totally different from @bygui86 usecase. For `templateRef` scenario can't refer the `spec` level properties. if you define the podmetadata under template it will work.

@sarabala1979 I didn't understand, when I read the spec  https://argoproj.github.io/argo-workflows/fields/#workflowtemplatespec 
there is a podmedata field for WorkflowTemplate. 
If I trigger the WotkflowTemplate with a templateRef from a Workflow or a CronWorkflow or argocli, I should see my podmetadata ?
Otherwise what the use of podmetadata for WorkflowTemplate ?

**Comment by alexec** [comment]
This issue is fixed. A new issue must be raised for @guillaumBrisard problem.

**Comment by alexec** [comment]
Reviewing this with @sarabala1979 , `templateRef` references a template with a workflow template, not the template. So anything not under the specific template will be ignored. This is as intended.

**Comment by bygui86** [comment]
@alexec probably I didn't fully understand your last comment...
As you can see in my initial message, `podMetadata` is declared in a `ClusterWorkflowTemplate` so it's part of a template and should not be ignored. But actually it is and only adding the same `podMetadata` declaration to the ArgoEvents Sensor I see labels and annotations on the workflow pods.

**Comment by guillaumBrisard** [comment]
@bygui86 
What I understood. If I use the @amaloun example.
To make it works you need to move `spec/podMetadata` from `WorkflowTemplate` to `Workflow`
Why ?
Because when we use `templateRef` in the `Workflow`, it will see what is under the field `spec/templates` in the WorkflowTemplate 
and everything not under `spec/templates` will be ignored like `podMetadata`,  `hostNetwork` or `dnsPolicy`.
So those fields need to be in the k8s object that trigger the `WorkflowTemplate` like a Workflow or a CronWorkflow.

In the end, I really don't understand what is the use of  `podMetadata`,  `hostNetwork` or `dnsPolicy` for a `WorkflowTemplate`
The doc should be updated because it's really confusing, if I'm correct

**Comment by bygui86** [comment]
@guillaumBrisard thanks for the explanation.

Yeah at this point I think that this behaviour is kind of wrong... I mean: overwriting what's in the WorkflowTemplate with what's in the Workflow, is fine. But if there is nothing in the Workflow, why ignoring what's in the template?

`podMetadata` are useful if you run your workflows in a cluster secured with NetworkPolicies (based on labelling).
`hostNetwork` and `dnsPolicy` are useful in a cluster secured with PSP and similar.
So I think that applying them only through argo cli is not enough.

> The doc should be updated because it's really confusing

This is anyway true...

**Comment by alexec** [comment]
It is confusing. We have “WorkflowTemplate”, “template” “workflowTemplateRef” and “templateRef”. I think is easy to misunderstand, so let me see if I can clear this up (or make it more confusing). 

You can’t inherit fields under spec with “templateRef”, because this refers to a specific “template” within the “WorkflowTemplate”, not the “WorkflowTemplate” itself. If that is what you are trying, then maybe the docs could be clearer.

You can (should) inherit if you use “spec.workflowTemplateRef”.  If that is what you are trying, then maybe there is a bug.

**Comment by guillaumBrisard** [comment]
@alexec Thanks
So `templateRef`, it's better if we want to compose an other `Workflow` from multiple `template` in multiple `WorkflowTemplate` 
But it's impossible to merge others fields than `template`. It make sense otherwise it will be impossible to understand

But if we want to just trigger a `WorkflowTemplate` from a `CronWorkflow`, it's better to use `spec.workflowTemplateRef`

In [Workflow Templates](https://argoproj.github.io/argo-workflows/workflow-templates/)
maybe it will be good to explain these use cases with their pros and cons

**Comment by bygui86** [comment]
@alexec thank you for the clarification.

It's really difficult to distinguish between following concepts, but probably I got them:
- `WorkflowTemplate`, CRD to declare a template for a workflow
- `workflowTemplateRef`, reference to a WorkflowTemplate CRD
- `template`, piece of a workflow defined inside a Workflow or WorkflowTemplate (spec.templates)
- `templateRef`, reference to a template

In the ArgoEvents Sensor I refer to a WorkflowTemplate using `workflowTemplateRef`. You can see it above in [my previous comment](https://github.com/argoproj/argo-workflows/issues/7091#issuecomment-955577224)
```yaml
...
                workflowTemplateRef:
                  name: go-pipeline
                  clusterScope: true
...
```
For this reason I think it could be a bug, but please correct me if I'm wrong!

**Comment by alexec** [comment]
@bygui86 I've failed to repro your issue. So maybe you have some special set-up. Could you set-up time via https://bit.ly/book-30m-with-argo-team so we can debug please?

**Comment by bygui86** [comment]
@alexec with pleasure, we can try to use my environment to see whether I made something weird.

Scheduled a call for the 16th Dec

**Comment by sarabala1979** [comment]
@bygui86  can you try on the new release? 3.1.14 or 3.2.4
@guillaumBrisard  can you create separate issue to discuss for `templateRef`?

**Comment by bygui86** [comment]
@sarabala1979 I will try in the next couple of weeks

**Comment by guillaumBrisard** [comment]
@sarabala1979 thanks, but it's fine for me now with the below explanation 

> It is confusing. We have “WorkflowTemplate”, “template” “workflowTemplateRef” and “templateRef”. I think is easy to misunderstand, so let me see if I can clear this up (or make it more confusing).
> 
> You can’t inherit fields under spec with “templateRef”, because this refers to a specific “template” within the “WorkflowTemplate”, not the “WorkflowTemplate” itself. If that is what you are trying, then maybe the docs could be clearer.
> 
> You can (should) inherit if you use “spec.workflowTemplateRef”. If that is what you are trying, then maybe there is a bug.

**Comment by guillaumBrisard** [comment]
@sarabala1979 thanks, but it's fine for me now with the below explanation 

> It is confusing. We have “WorkflowTemplate”, “template” “workflowTemplateRef” and “templateRef”. I think is easy to misunderstand, so let me see if I can clear this up (or make it more confusing).
> 
> You can’t inherit fields under spec with “templateRef”, because this refers to a specific “template” within the “WorkflowTemplate”, not the “WorkflowTemplate” itself. If that is what you are trying, then maybe the docs could be clearer.
> 
> You can (should) inherit if you use “spec.workflowTemplateRef”. If that is what you are trying, then maybe there is a bug.

**Comment by alexec** [comment]
We tested today. We found that this is a bug in v3.1.3 is fixed in v3.2.4.
